### ISSUE TITLE
docker.login module is failing

### ISSUE THREAD

**Issue by dennisoconnor** [saltstack/salt#18673]
The docker.login module command is not functioning. This is also affecting docker.pulled states from private repositories.

Docker version 1.3.1, build 4e9bbfa
salt 2014.1.13 (Hydrogen)
docker-py (0.6.0)

```
$ sudo salt-call docker.login https://index.docker.io/v1/ test@email.com testpassword testusername

[ERROR   ] An un-handled exception was caught by salt's global exception handler:
APIError: 500 Server Error: Internal Server Error ("Invalid registry endpoint https://testusername/v1/: Get https://testusername/v1/_ping: dial tcp: i/o timeout. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry testusername` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/testusername/ca.crt")
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 82, in salt_call
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/__init__.py", line 319, in run
    caller.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 144, in run
    ret = self.call()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 81, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/modules/dockerio.py", line 1167, in login
    return client.login(url, username, password, email)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 739, in login
    return self._result(response, json=True)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 93, in _result
    self._raise_for_status(response)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 89, in _raise_for_status
    raise errors.APIError(e, response, explanation=explanation)
APIError: 500 Server Error: Internal Server Error ("Invalid registry endpoint https://testusername/v1/: Get https://testusername/v1/_ping: dial tcp: i/o timeout. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry testusername` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/testusername/ca.crt")
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 82, in salt_call
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/__init__.py", line 319, in run
    caller.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 144, in run
    ret = self.call()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 81, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/modules/dockerio.py", line 1167, in login
    return client.login(url, username, password, email)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 739, in login
    return self._result(response, json=True)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 93, in _result
    self._raise_for_status(response)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 89, in _raise_for_status
    raise errors.APIError(e, response, explanation=explanation)
docker.errors.APIError: 500 Server Error: Internal Server Error ("Invalid registry endpoint https://testusername/v1/: Get https://testusername/v1/_ping: dial tcp: i/o timeout. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry testusername` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/testusername/ca.crt")
```

**Comment by colincoghill** [comment]
Not sure if it's the same problem, but it's also failing for me with 2014.7.0

salt-call 2014.7.0 (Helium)
docker-py   0.7.1
docker  1.4.1
Ubuntu Trusty 

I get the same error to https://quay.io/v1/ also. (with valid credentials)



$ sudo salt-call docker.login https://index.docker.io/v1/ test@email.com testpassword testusername[WARNING ] /usr/lib/python2.7/dist-packages/salt/modules/config.py:251: DeprecationWarning: pillar_opts will default to False in the Lithium release
  'pillar_opts will default to False in the Lithium release'

[INFO    ] Starting new HTTPS connection (1): testusername
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
DockerException: HTTPS endpoint unresponsive and insecure mode isn't enabled.
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 122, in salt_call
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/__init__.py", line 419, in run
    caller.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 225, in run
    ret = self.call()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 129, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/modules/dockerio.py", line 1151, in login
    return client.login(url, username, password, email)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 739, in login
    if registry else auth.INDEX_URL
  File "/usr/local/lib/python2.7/dist-packages/docker/auth/auth.py", line 46, in expand_registry_url
    "HTTPS endpoint unresponsive and insecure mode isn't enabled."
DockerException: HTTPS endpoint unresponsive and insecure mode isn't enabled.
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 122, in salt_call
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/__init__.py", line 419, in run
    caller.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 225, in run
    ret = self.call()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 129, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/modules/dockerio.py", line 1151, in login
    return client.login(url, username, password, email)
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 739, in login
    if registry else auth.INDEX_URL
  File "/usr/local/lib/python2.7/dist-packages/docker/auth/auth.py", line 46, in expand_registry_url
    "HTTPS endpoint unresponsive and insecure mode isn't enabled."
docker.errors.DockerException: HTTPS endpoint unresponsive and insecure mode isn't enabled.

**Comment by rallytime** [comment]
Thanks for the confirmation @colincoghill. We'll take a look.

**Comment by colincoghill** [comment]
In salt/modules/dockerio.py, ~ line 1151:
    client = _get_client()
    return client.login(url, username, password, email)

It appears to call docker-py's:

    def login(self, username, password=None, email=None, registry=None,
              reauth=False, insecure_registry=False, dockercfg_path=None):

Which takes the arguments in a different order. This could explain part of the problem?  I've corrected that one line, but it still fails later :)

**Comment by colincoghill** [comment]
Actually the above fixes it for me if I call:
sudo salt-call docker.login https://index.docker.io/v1/ testuser testpassword testemail

so just that fix on ~1151 in dockerio.py

is that enough or shall I submit a pull request? (it's just one line)

**Comment by colincoghill** [comment]
Made a PR:  https://github.com/saltstack/salt/pull/19855

**Comment by rallytime** [comment]
Thanks for the PR! It looks like this is something that should be fixed on the 2014.7 branch as well, so I'll label your PR for backporting.
